webhooks/clusterclass.go

```diff
@@ -123,8 +140,126 @@ func (webhook *ClusterClass) validate(old, new *clusterv1.ClusterClass) error {
 	// Ensure spec changes are compatible.
 	allErrs = append(allErrs, check.ClusterClassesAreCompatible(old, new)...)
 
+	// Ensure no MachineDeploymentClass currently in use has been removed from the ClusterClass.
+	allErrs = append(allErrs, webhook.validateLivingMachineDeploymentClassesNotRemoved(old, new)...)
+
 	if len(allErrs) > 0 {
 		return apierrors.NewInvalid(clusterv1.GroupVersion.WithKind("ClusterClass").GroupKind(), new.Name, allErrs)
 	}
 	return nil
 }
+
+func (webhook *ClusterClass) validateLivingMachineDeploymentClassesNotRemoved(old, new *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+	if old == nil {
+		return allErrs
+	}
+	missingClasses := webhook.removedMachineClasses(old, new)
+
+	// If no classes have been removed return early as no further checks are needed.
+	if len(missingClasses) == 0 {
+		return allErrs
+	}
+
+	// TODO: The first part of this method will need to be broken out to properly account for changes when also validating variables.
+	// Retrieve all clusters using the ClusterClass.
+	clusters, errs := webhook.clustersUsingClusterClass(old)
+	if len(errs) > 0 {
+		allErrs = append(allErrs, errs...)
+		return allErrs
+	}
+
+	// Retrieve all the MachineDeployments in the clusterClass namespace using a managed topology.
+	machineDeployments, errs := webhook.topologyManagedMachineDeployments(new.Namespace)
+	if len(errs) > 0 {
+		allErrs = append(allErrs, errs...)
+		return allErrs
+	}
+
+	// Create a set of records of machineDeploymentClass information where the name is a fully qualified name for the
+	// MachineDeployment topology in the form cluster.Name/machineDeploymentTopology.Name and links to the machineDeploymentClass Name.
+	mdcRecords := map[string]string{}
+	for _, c := range clusters {
+		for _, machineDeploymentClass := range c.Spec.Topology.Workers.MachineDeployments {
```

Comment:
```suggestion
		for _, machineDeploymentTopology := range c.Spec.Topology.Workers.MachineDeployments {
```
nit, or mdTopology

Reply:
There's a tradeoff here on which race conditions we want to account for I think. If we just check the cluster topologies there may be some impact on existing machineDeployments on being reconciled before deletion.

Just checking the cluster object makes the code a lot simpler and more efficient though.